example/libclang-example/generated_bindings.dart

```diff
@@ -9282,7 +9283,7 @@ typedef DartClang_Cursor_getTranslationUnit = CXTranslationUnit Function(
 class CXCursorSetImpl extends ffi.Opaque {}
```

Comment:
On second thought, I am not sure if we shouldn't generate the custom import.

The user has control over what declaration to include/exclude. Currently, if the user has excluded a declaration, we will not generate it. The custom import will be used instead.

But if we force exclude anything that is inside a custom import. We will never generate it even if the user specifies it.

Reply:
I was thinking that we either:

1. include it from a custom import
2. or generate it

But don't do both and generate an unused declaration.

> Currently, if the user has excluded a declaration, we will not generate it.
In other words: currently, if the user does not exclude something but includes a header file, we generate it.

And currently, we only generate declarations that are (transitively) reachable from other files. We don't generate things that are not used at all.

I can see how these things conflict.

We need to come up with a reasonable way to handle 'shadowing'.

I'd say that shadowing would be: if we have already defined a type in an import, do not regenerate it, unless it is specifically included with an include.

(I just realized that us having imports will also start allowing us to generate bindings that import other generated bindings. 🎉  And it would be also relevant for that. In that case it might even refer to the same sources in C.)

What do you think?

Reply:
How about adding a custom key for shadow, with default as true. Which can be overriden.
```yaml
type-map:
  'struct MyStruct':
    'lib': 'custom_lib'
    'c-type': 'CustomType'
    'dart-type': 'CustomType'
    'shadow': false
```


Generated bindings using other generated bindings is exactly what I had in mind 🎉

Reply:
> How about adding a custom key for shadow

Hm, if the key is set to true, then the declarations are not generated, but that is not visible in the part where you specify your includes/excludes on those declarations.
And, if the key is set to false, then you might get weird results when the file you're having a custom import in changes and now defines it. You might not recognize that it's shadowed.

I was thinking more of a general shadowing behavior rather than per item:

`type-map-shadowing: prefer-type-map (default) | prefer-local | error`.

With `error`: let the users resolve it with the include/exclude of the main file and the list of imports through the type-map.
With `prefer-type-map`: the type-map always shadows local definitions and we don't generate local definitions
With `prefer-local`: local definitions win out. (Although, I cannot come up with a use case for this.)

Though, this line of thinking does not distinguish between declarations in the `entry-points` and `include-directives`.

I think the default behavior for a conflict between the `type-map` and non-excluded definitions in `entry-points` should be `error`. While the default behavior for a conflict between the `type-map` and `include-directives` should be that the `type-map` wins.

If we have:

* `a.h` defining `struct A`
* `b.h` including `a.h` and using `struct A` and defining a `struct B`.
* `c.h` including `b.h` and defining a function with `struct B`.
* `d.h` also defining `struct A` in the same way as `a.h`.

And then we have two `ffigen` configs:

1. generating for `b.h` with `a.h` in its include directives. (and outputting also a yaml file with a type-map? 😄 )
2. generating for `c.h` with both other files in its include directives and having a type-map for `struct A` and `struct B`.
3. generating for `d.h` with type-map for `struct A`

In (1) we generate `struct A` because it is not defined in `b.h`.
In (2) we don't generate `struct A` and `B` because it's already in the `type-map`.
In (3) we get an error, because we would get conflicting definitions. The user has to not include the local `struct A`.

If we go this route we don't need a config key at all, but some documentation! 😸 

Do you envision use cases in which this would not work?

Reply:
This honesty seems too complicated to me. Infact there's no good reason for us to error out and make the user exclude declarations.
The only use case I see for prefer-local would be to sort of compare the generated bindings to see if the struct definitions are same, and then remove them.
Perhaps we can just have a global key to exclude type-map declarations from being generated. A  type-map level key like shadow (from earlier suggestion) would be too verbose.

Also the entry-points are only a starting point for parsing. Whether or not we include a declaration depends on what matched by include-directives.

Reply:
> Also the entry-points are only a starting point for parsing. Whether or not we include a declaration depends on what matched by include-directives.

Ah, I misremembered that then. Sorry for the confusion.

Let's just implement something for now, so that we can have support for the `type-map`. And then we can always refine that (with a breaking change and major version bump) if we run into use cases which we didn't foresee.